src/test/objectstore/store_test.cc

```diff
@@ -7288,6 +7294,55 @@ TEST_P(StoreTestSpecificAUSize, ZeroBlockDetectionSmall) {
     ASSERT_EQ(logger->get(l_bluestore_write_small), 2u);
     ASSERT_EQ(logger->get(l_bluestore_write_small_bytes), 4096u*2);
     ASSERT_EQ(logger->get(l_bluestore_zero_block_count_small), 1u);
+
+    // [4] read-after-non-zero-write check
+    bufferlist in, _exp;
+    r = store->read(ch, hoid, 0, 0x4000, in);
+    ASSERT_EQ(4096, r);
+    ASSERT_FALSE(in.is_zero());
+    _exp.append(bl);
+    ASSERT_TRUE(bl_eq(_exp, in));
+  }
+
+  {
+    // [5] zero-overwrite check
+    ObjectStore::Transaction t;
+    bufferlist bl;
+
+    bl.append(std::string(4096, 'c'));
+    bl.zero(0, 2048); // partially overwrite
```

Comment:
Sorry I meant a bit different overwrite scenario. 
Appending write means that  writing extends object size and written block doesn't invalidate any existing data.
Contrary overwrite one doesn't change object size and invalidates  existing block. 
So at [1] you append zeros and at [3] overwrite data with non-zero block. What's missing is overwriting non-zero block with zero one.

Some more versatile write patterns would be nice to have too, e.g. mixing small/big and zero/non-zero writes (e.g. overwriting 'big' non zero block with small zero one or vice verso),  aligned/unaligned small zero blocks, big zero blocks longer than 4K  etc... But that's rather optional, up to you to add that...

And please note that overwrites (as per above specification) aren't currently supported by smr drives, appends  only - hence that is_smr check isn't absolutely useless. You might want either selectively bypass overwrite subcases  or have simple append-only test case(s) supporting SMR and more versatile ones which are omitted in case of SMRs.

Reply:
@ifed01 I see what you mean here-- I misunderstood the original prompt. I've been working on some new changes that I'll push here, and I'll re-request your review when I'm sure all the tests are passing.